rgw: add support for FormPost of Swift API

rzarzynski opened the pull request:
Both tempest and s3-tests haven't found any regression here.

CC: @yehudasa, @cbodley

rzarzynski commented:
Rebased due to a merge conflict.

rzarzynski commented:
@yehudasa, @mattbenjamin, @cbodley: rebased due to merge conflicts after getting in the compression stuff and front-end rework.

I will retest with Tempest and s3-tests.

rzarzynski commented:
Retested. Tempest and s3-tests haven't found any new issue on this branch in comparison to `master`. However, the fresh `master` seems to have a regression related to Swift's object versioning (created [bug #17803](http://tracker.ceph.com/issues/17803)).

mattbenjamin commented:
@rzarzynski could you rebase?  tx!

rzarzynski commented:
It looks the PR got new merge conflicts yesterday. Rebasing and testing.

rzarzynski commented:
@mattbenjamin: both s3-tests and Tempest haven't found any regression in comparison to the current `master` (3a9bcaa4aa6042587886c0eaae0ce4eeeb8f8164, `Fri Dec 2 14:48:15 2016 +0800`).

s3-tests found an unrelated issue in `master` (so no regression here) which causes a failure in following tests: 
```
s3tests.functional.test_s3.test_versioning_multi_object_delete
s3tests.functional.test_s3.test_versioning_multi_object_delete_with_marker
```
I'm going to fill a ticket for that.

Tempest confirmed the FormPost is operational:
```
$ ./run_tempest.sh -V tempest.api.object_storage.test_object_formpost
...
{1} tempest.api.object_storage.test_object_formpost.ObjectFormPostTest.test_post_object_using_form [0.048283s] ... ok
{0} tempest.api.object_storage.test_object_formpost_negative.ObjectFormPostNegativeTest.test_post_object_using_form_expired [2.028040s] ... ok
{0} tempest.api.object_storage.test_object_formpost_negative.ObjectFormPostNegativeTest.test_post_object_using_form_invalid_signature [0.011914s] ... ok

======
Totals
======
Ran: 3 tests in 5.0000 sec.
 - Passed: 3
 - Skipped: 0
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 2.0882 sec.

==============
Worker Balance
==============
 - Worker 0 (2 tests) => 0:00:02.042187
 - Worker 1 (1 tests) => 0:00:00.048283
```

I appended a7fead41cf25352d93e69a3f957fb33612aeb658 which slightly modifies error handling in the situation when a form is expired or has an invalid signature. It looks that in the meanwhile Tempest became more restrictive. Previously both `401 Unauthorized` and `403 Forbidden` were acceptable. After upgrading to a newer version (`51feb121085cf0a4d70af4b2b086f03991d5bceb`, `Wed Nov 16 10:13:27 2016 +0000`) only the first one is seen as correct.

rzarzynski commented:
@mattbenjamin: according to [the Casey's comment](https://github.com/ceph/ceph/pull/12283#issuecomment-264522550) the fix for the issue in `master` has been merged today.

oritwas commented:
@rzarzynski please rebase

rzarzynski commented:
@oritwas, @mattbenjamin: rebased to `master`. Tempest nor s3-tests haven't found any regression here. In a dedicated run Tempest additionally confirmed the FormPost is operational:
```
$ ./run_tempest.sh -V tempest.api.object_storage.test_object_formpost
WARNING: This script is deprecated and will be removed in the near future. Please migrate to tempest run or another method of launching a test runner
running=OS_STDOUT_CAPTURE=${OS_STDOUT_CAPTURE:-1} \
OS_STDERR_CAPTURE=${OS_STDERR_CAPTURE:-1} \
OS_TEST_TIMEOUT=${OS_TEST_TIMEOUT:-500} \
OS_TEST_LOCK_PATH=${OS_TEST_LOCK_PATH:-${TMPDIR:-'/tmp'}} \
${PYTHON:-python} -m subunit.run discover -t ${OS_TOP_LEVEL:-./} ${OS_TEST_PATH:-./tempest/test_discover} --list 
running=OS_STDOUT_CAPTURE=${OS_STDOUT_CAPTURE:-1} \
OS_STDERR_CAPTURE=${OS_STDERR_CAPTURE:-1} \
OS_TEST_TIMEOUT=${OS_TEST_TIMEOUT:-500} \
OS_TEST_LOCK_PATH=${OS_TEST_LOCK_PATH:-${TMPDIR:-'/tmp'}} \
${PYTHON:-python} -m subunit.run discover -t ${OS_TOP_LEVEL:-./} ${OS_TEST_PATH:-./tempest/test_discover}  --load-list /tmp/tmpVXDbSs
running=OS_STDOUT_CAPTURE=${OS_STDOUT_CAPTURE:-1} \
OS_STDERR_CAPTURE=${OS_STDERR_CAPTURE:-1} \
OS_TEST_TIMEOUT=${OS_TEST_TIMEOUT:-500} \
OS_TEST_LOCK_PATH=${OS_TEST_LOCK_PATH:-${TMPDIR:-'/tmp'}} \
${PYTHON:-python} -m subunit.run discover -t ${OS_TOP_LEVEL:-./} ${OS_TEST_PATH:-./tempest/test_discover}  --load-list /tmp/tmpDY58PT
{1} tempest.api.object_storage.test_object_formpost.ObjectFormPostTest.test_post_object_using_form [0.073423s] ... ok
{0} tempest.api.object_storage.test_object_formpost_negative.ObjectFormPostNegativeTest.test_post_object_using_form_expired [2.022728s] ... ok
{0} tempest.api.object_storage.test_object_formpost_negative.ObjectFormPostNegativeTest.test_post_object_using_form_invalid_signature [0.012430s] ... ok

======
Totals
======
Ran: 3 tests in 7.0000 sec.
 - Passed: 3
 - Skipped: 0
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 2.1086 sec.

==============
Worker Balance
==============
 - Worker 0 (2 tests) => 0:00:02.035737
 - Worker 1 (1 tests) => 0:00:00.073423
```

The updated branch contains the change [requested  by @oritwas](https://github.com/ceph/ceph/pull/11066#discussion_r109618925) in PR #11066.

rzarzynski commented:
@oritwas: the branch was a subject of following Teuthology runs:
* http://pulpito.ceph.com/rzarzynski-2017-04-07_12:57:58-rgw-wip-rzarzynski-testing---basic-mira/ (the main one; 30 jobs failed due to a problem with a package fetch),
* http://pulpito.ceph.com/rzarzynski-2017-04-09_14:23:47-rgw-wip-rzarzynski-testing---basic-mira/ (rerun of failed/dead jobs from the previous one).

Finally we have 6 unrelated failures. All of them are because of the well-known Valgrind issue:
`strange leak of std::string memory from md_config_t seen in radosgw`.

oritwas commented:
@rzarzynski , needs another rebase

oritwas commented on src/rgw/rgw_rest.cc:
```diff
@@ -1235,6 +1235,306 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
+ */
+static int index_of(bufferlist& bl, int max_len, const string& str,
+		    bool check_crlf,
+		    bool *reached_boundary, int *skip)
```
I prefer reference for reached_boundary and skip otherwise you need to check they are not null

oritwas commented on src/rgw/rgw_rest.cc:
```diff
@@ -1235,6 +1235,306 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
+int RGWPostObj_ObjStore::read_with_boundary(ceph::bufferlist& bl,
+                                            uint64_t max,
+                                            const bool check_crlf,
+                                            bool* const reached_boundary,
```
reference ...

oritwas commented on src/rgw/rgw_rest.cc:
```diff
@@ -1235,6 +1235,306 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
+                                            uint64_t max,
+                                            const bool check_crlf,
+                                            bool* const reached_boundary,
+                                            bool* const done)
```
here too

oritwas commented on src/rgw/rgw_rest.cc:
```diff
@@ -1235,6 +1235,306 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
+  return true;
+}
+
+std::string RGWPostObj_ObjStore::get_part_str(/*const*/ parts_coll_t& parts,
```
remove the comment or add the const if you can

oritwas commented on src/rgw/rgw_rest.cc:
```diff
@@ -1235,6 +1235,306 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
+  }
+}
+
+bool RGWPostObj_ObjStore::part_bl(/*const*/ parts_coll_t& parts,
```
same here

oritwas commented on src/rgw/rgw_rest.h:
```diff
@@ -216,15 +216,78 @@ class RGWPutObj_ObjStore : public RGWPutObj
   int get_padding_last_aws4_chunk_encoded(bufferlist &bl, uint64_t chunk_size);
 };
 
+
```
nit: empty line

oritwas commented on src/rgw/rgw_rest.h:
```diff
@@ -216,15 +216,78 @@ class RGWPutObj_ObjStore : public RGWPutObj
+    ceph::bufferlist data;
+  };
+
+  using parts_coll_t = \
```
don't like this name parts_collection_t maybe?

oritwas commented on src/rgw/rgw_rest.h:
```diff
@@ -216,15 +216,78 @@ class RGWPutObj_ObjStore : public RGWPutObj
+                                    std::string& first,
+                                    std::map<std::string, std::string>& params);
+
+  static bool part_str(/*const*/ parts_coll_t& parts,
```
const comment

oritwas commented on src/rgw/rgw_rest.h:
```diff
@@ -216,15 +216,78 @@ class RGWPutObj_ObjStore : public RGWPutObj
+                       const std::string& name,
+                       std::string *val);
+
+  static std::string get_part_str(/*const*/ parts_coll_t& parts,
```
and here

oritwas commented on src/rgw/rgw_rest.h:
```diff
@@ -216,15 +216,78 @@ class RGWPutObj_ObjStore : public RGWPutObj
+                                  const std::string& name,
+                                  const std::string& def_val = std::string());
+
+  static bool part_bl(/*const*/ parts_coll_t& parts,
```
and here

rzarzynski commented:
@oritwas: thanks for the review! I've addressed your comments, rebased the branch and verified it locally. Tempest nor s3-tests (both in AWSv2 and AWSv4 mode) haven't found any regression here. Additionally the Tempest confirmed the FormPost is operational (a separated run):
```
$ ./run_tempest.sh -V tempest.api.object_storage.test_object_formpost
WARNING: This script is deprecated and will be removed in the near future. Please migrate to tempest run or another method of launching a test runner
running=OS_STDOUT_CAPTURE=${OS_STDOUT_CAPTURE:-1} \
OS_STDERR_CAPTURE=${OS_STDERR_CAPTURE:-1} \
OS_TEST_TIMEOUT=${OS_TEST_TIMEOUT:-500} \
OS_TEST_LOCK_PATH=${OS_TEST_LOCK_PATH:-${TMPDIR:-'/tmp'}} \
${PYTHON:-python} -m subunit.run discover -t ${OS_TOP_LEVEL:-./} ${OS_TEST_PATH:-./tempest/test_discover} --list 
running=OS_STDOUT_CAPTURE=${OS_STDOUT_CAPTURE:-1} \
OS_STDERR_CAPTURE=${OS_STDERR_CAPTURE:-1} \
OS_TEST_TIMEOUT=${OS_TEST_TIMEOUT:-500} \
OS_TEST_LOCK_PATH=${OS_TEST_LOCK_PATH:-${TMPDIR:-'/tmp'}} \
${PYTHON:-python} -m subunit.run discover -t ${OS_TOP_LEVEL:-./} ${OS_TEST_PATH:-./tempest/test_discover}  --load-list /tmp/tmpHoSEVb
running=OS_STDOUT_CAPTURE=${OS_STDOUT_CAPTURE:-1} \
OS_STDERR_CAPTURE=${OS_STDERR_CAPTURE:-1} \
OS_TEST_TIMEOUT=${OS_TEST_TIMEOUT:-500} \
OS_TEST_LOCK_PATH=${OS_TEST_LOCK_PATH:-${TMPDIR:-'/tmp'}} \
${PYTHON:-python} -m subunit.run discover -t ${OS_TOP_LEVEL:-./} ${OS_TEST_PATH:-./tempest/test_discover}  --load-list /tmp/tmpE2P2gE
{1} tempest.api.object_storage.test_object_formpost.ObjectFormPostTest.test_post_object_using_form [0.045707s] ... ok
{0} tempest.api.object_storage.test_object_formpost_negative.ObjectFormPostNegativeTest.test_post_object_using_form_expired [2.024572s] ... ok
{0} tempest.api.object_storage.test_object_formpost_negative.ObjectFormPostNegativeTest.test_post_object_using_form_invalid_signature [0.010969s] ... ok

======
Totals
======
Ran: 3 tests in 5.0000 sec.
 - Passed: 3
 - Skipped: 0
 - Expected Fail: 0
 - Unexpected Success: 0
 - Failed: 0
Sum of execute time for each test: 2.0812 sec.

==============
Worker Balance
==============
 - Worker 0 (2 tests) => 0:00:02.035966
 - Worker 1 (1 tests) => 0:00:00.045707
```

oritwas commented on src/rgw/rgw_rest.cc:
```diff
@@ -1236,6 +1236,308 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
+void RGWPostObj_ObjStore::parse_boundary_params(const std::string& params_str,
+                                                std::string& first,
+                                                std::map<std::string,
+                                                         std::string>& params)
```
nit: too much spacing
